src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -330,64 +278,76 @@
    handler-fn :- message-fn-schema]
   (swap! listener-atom conj [pred handler-fn]))
 
-(defn start-receiver
-  [connection endpoint discard-dir process-msg]
-  (let [sess (.createSession connection true 0)
-        q (.createQueue sess endpoint)
-        consumer (.createConsumer sess q)
-        producer (.createProducer sess q)
-        send #(mq/commit-or-rollback sess
-                (.send producer (mq/to-jms-message sess %1 %2)))
-        handle (create-message-handler send discard-dir process-msg)]
-    (.setMessageListener
-     consumer
-     (reify MessageListener
-       (onMessage [this msg]
-         (try
-           (mq/commit-or-rollback sess (handle (mq/convert-jms-message msg)))
-           (catch Throwable ex
-             (log/error ex "message receive failed")
-             (throw ex))))))
-    {:session sess :consumer consumer :producer producer}))
+(def one-hour (* 1000 60 60))
+
+(defn send-delayed-message [command-chan delay-pool]
+  (fn [cmd]
+    (let [narrowed-entry (dissoc cmd :payload)]
+      (after one-hour #(async/>!! command-chan narrowed-entry) delay-pool))))
```

Comment:
Previously failed commands were enqueued to the scheduler queue, that was periodically polled, then re-enqueued in the regular queue. This behavior is now handled in memory. `delay-pool` is a Java scheduled threadpool, managed by atat. Failures require no disk writes. This function removes the payload to consume less memory. That command ref stays in the delay-pool for one hour, then invokes the above function, which re-enqueues the command ref.